flink-test-utils-parent/flink-connector-testing/src/main/java/org/apache/flink/connectors/test/common/utils/TestDataMatchers.java

```diff
@@ -151,11 +156,15 @@ protected boolean matchesSafely(Iterator<T> resultIterator, Description descript
                 final T record = resultIterator.next();
                 if (!matchThenNext(record)) {
                     mismatchDescription = String.format("Unexpected record '%s'", record);
+                    logCurrentProgress();
                     return false;
                 }
             }
             if (!hasReachedEnd()) {
-                mismatchDescription = "Result data is less than test data";
+                mismatchDescription =
+                        "Number of records consumed from external system is less than "
+                                + "records written to external system";
```

Comment:
This message apparently was also not very helpful. Could we add how many matched and which record is the first that didn't?